benchmark/test/test_script.py

```diff
@@ -424,3 +427,101 @@ def test_default_metric(self):
         """
         config = get_config_by_name(self.config['metrics'], 'default')
         self.assertEqual(config['name'], 'default')
+
+
+class UserDataTests(TestCase):
+    """
+    Test --userdata option.
+    """
+
+    def test_no_userdata(self):
+        """
+        Missing option adds nothing to result.
+        """
+        options = BenchmarkOptions()
+        options.parseOptions([])
+        result = {}
+        add_userdata(options, result)
+        self.assertEqual(result, {})
+
+    def test_empty_userdata(self):
+        """
+        Empty option adds nothing to result.
+        """
+        options = BenchmarkOptions()
+        options.parseOptions(['--userdata', ''])
+        result = {}
+        add_userdata(options, result)
+        self.assertEqual(result, {})
+
+    def test_json_userdata(self):
+        """
+        JSON string adds to result.
+        """
+        options = BenchmarkOptions()
+        options.parseOptions(['--userdata', '{"branch": "master"}'])
+        result = {}
+        add_userdata(options, result)
+        self.assertEqual(result, {"userdata": {"branch": "master"}})
+
+    def test_json_file_userdata(self):
+        """
+        JSON file adds to result.
+        """
+        json_file = FilePath(self.mktemp())
+        with json_file.open('w') as f:
+            f.write('{"branch": "master"}\n')
+        options = BenchmarkOptions()
+        options.parseOptions(['--userdata', '@{}'.format(json_file.path)])
+        result = {}
+        add_userdata(options, result)
+        self.assertEqual(result, {"userdata": {"branch": "master"}})
+
+    def test_invalid_json(self):
+        """
+        Invalid JSON string handled.
+        """
+        options = BenchmarkOptions()
+        options.parseOptions(['--userdata', '"branch": "master"'])
+        with capture_stderr() as captured_stderr:
```

Comment:
:question: It's introduced in this branch, so optional, but I think `capture_stderr` needs to have a `try` / `finally` around the yield. Probably there should also be tests to check that it restores stderr.